File: 백준/Silver/14501. 퇴사/퇴사.java
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int n = Integer.parseInt(br.readLine());
        int[] t = new int[n];
        int[] p = new int[n];
        int[] dp = new int[n + 1];

        for (int i = 0; i < n; i++) {
            String[] input = br.readLine().split(" ");
            t[i] = Integer.parseInt(input[0]);
            p[i] = Integer.parseInt(input[1]);
        }

        for (int i = n - 1; i >= 0; i--) {
            int period = i + t[i];

            if (period <= n) {
                dp[i] = Math.max(p[i] + dp[period], dp[i + 1]);
            } else {
                dp[i] = dp[i + 1];
            }
        }

        System.out.println(dp[0]);
    }
}

File: 백준/Silver/14501. 퇴사/퇴사.js
let [n, ...input] = require("fs")
  .readFileSync(process.platform === "linux" ? "/dev/stdin" : "input.txt")
  .toString()
  .trim()
  .split("\n");

n = Number(n);
input = input.map((v) => v.split(" ").map(Number));
let dp = Array(n + 1).fill(0);

for (let i = n - 1; i >= 0; i--) {
  let temp = i + Number(input[i][0]);

  if (temp <= n) {
    dp[i] = Math.max(input[i][1] + dp[temp], dp[i + 1]);
  } else {
    dp[i] = dp[i + 1];
  }
}

console.log(dp[0]);
